### PR TITLE
OCPBUGS-6829: Add support for protocolStrategy API field to enable force_tcp configuration

### PR DESCRIPTION
**Pull request by gcs278**
Support for `protocolStrategy` API field to CoreDNS Corefile that allows `force_tcp` to be enabled on any upstream server configuration.

API: https://github.com/openshift/api/pull/1429

Adds support for the following fields:
- `spec.upstreamResolvers.protocolStrategy`
  - For default (".") usptream server
- `spec.servers[].forwardPlugin.protocolStrategy`
  - For custom list of DNS resolvers

### CODE REVIEW

**Comment by gcs278** [comment]
/hold
requires API bump https://github.com/openshift/api/pull/1429

**Review by alebedev87 — COMMENTED** [review]
Logic LGTM, will wait for the API bump to revisit. Just a small question.

File: pkg/operator/controller/controller_dns_configmap_test.go [hunk 1]
```diff
@@ -908,6 +908,88 @@ func TestDesiredDNSConfigmapUpstreamResolvers(t *testing.T) {
...
+				},
+				Spec: operatorv1.DNSSpec{
+					Servers: []operatorv1.Server{
+						{
+							Name:  "foo",
+							Zones: []string{"foo.com"},
+							ForwardPlugin: operatorv1.ForwardPlugin{
+								Upstreams:      []string{"1.1.1.1", "2.2.2.2:5353"},
+								Policy:         operatorv1.RoundRobinForwardingPolicy,
+								ProtocolPolicy: operatorv1.ProtocolPolicyNone,
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply pkg/operator/controller/controller_dns_configmap_test.go:947]: _Update_: I think that makes some sense. Sort of double check on the defaulting or `omitempty` option. This way we have all the three possibilities covered: nothing is specified by the user, none and forcetcp. Never mind my comment.

**Comment by melvinjoseph86** [comment]
Pre verify the changes with cluster bot and steps are below:
https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-63512 
/label qe-approved

File: pkg/operator/controller/controller_dns_configmap_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (gcs278) [reply pkg/operator/controller/controller_dns_configmap_test.go]: With recent updates to this API, which have just merged, we now have `""` and `TCP`. `omitempty` was removed due to new API conventions (or previously un-followed conventions). 
https://github.com/openshift/api/blob/333bc194ef7a13338a4343c25f81ff856ab0d0fd/operator/v1/types_dns.go#L425

**Comment by gcs278** [comment]
We need to bump k8s to 0.27, but we are currently broke on https://github.com/openshift/cluster-dns-operator/pull/368#issuecomment-1551676822.

Using https://github.com/openshift/cluster-dns-operator/pull/368 to bump k8s.

**Comment by gcs278** [comment]
/unhold
1.27 rebase has merged. Rebased and ready to go.

**Comment by Miciah** [comment]
/approve
/lgtm
/hold
for @alebedev87.

**Pull request merged**